### PR TITLE
OAK-10225 Utility to rate limit writes in case async indexing is delayed

### PR DESCRIPTION
**Pull request by thomasmueller**

### CODE REVIEW

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import org.apache.jackrabbit.util.ISO8601;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RateLimitUtils {
+    
+    private final static Logger LOG = LoggerFactory.getLogger(RateLimitUtils.class);    
+
+    // when getOldestAsyncIndexUpdate was called the last time 
+    private static Instant lastJmxBeanCall = Instant.now();
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java:37]: this is never set again. Can it be marked as final?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    // the return value of the last call to getOldestAsyncIndexUpdate
+    private static Instant cachedOldestAsyncUpdate = Instant.now();
+    
+    // the last time the rateLimitWrites was called
+    private static Instant lastRateLimitCall = Instant.now();
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java:43]: Should all these static vars be uppercased?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    }
+    
+    /**
+     * Get the timestamp of the oldest indexing lane. In the normal case, this is at
+     * most 5 seconds ago, if indexing lanes are updated every 5 seconds. If
+     * indexing is delayed, or paused, this can be in the past.
+     * 
+     * The method can be called often without issues. If the last call was less than
+     * one second ago, the last returned value is returned. This is to avoid
+     * unnecessary calls to to the JMX beans.
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java]: typo
```suggestion
     * unnecessary calls to the JMX beans.
```

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java (hunk 2 above)
Comment (nit0906) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java:43]: +1

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java (hunk 1 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java:37]: Ah I forgot to set it!

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java (hunk 2 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/RateLimitUtils.java:43]: They are not constants, they are only static... According to the [Google Java Style](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names), only constants should be all uppercase. But I see it depends on who you ask: https://stackoverflow.com/questions/1417190/should-a-static-final-logger-be-declared-in-upper-case

**Review by fabriziofortino — APPROVED** [review]

**Pull request merged**